File: QUESTION2/question.py
programs = [
    {"name": "News", "duration": 1, "type": "news", "reach": 9, "revenue": 1000, "rights": (0, 24)},
    {"name": "Movie", "duration": 2, "type": "entertainment", "reach": 10, "revenue": 5000, "rights": (18, 22)},
    {"name": "Kids Show", "duration": 1, "type": "kids", "reach": 6, "revenue": 800, "rights": (8, 18)},
]

# 24 time slots (1 hour per slot for simplicity)
slots = [None] * 24
prime_time = range(18, 22)

# Sort programs by reach * revenue (weight)
programs.sort(key=lambda p: (p["reach"] * p["revenue"]), reverse=True)

for program in programs:
    start, end = program["rights"]
    for i in range(start, end - program["duration"] + 1):
        if all(slots[j] is None for j in range(i, i + program["duration"])):
            if any(j in prime_time for j in range(i, i + program["duration"])):
                for j in range(i, i + program["duration"]):
                    slots[j] = program["name"]
                break

# Print Schedule
for hour, program in enumerate(slots):
    print(f"{hour}:00 - {hour+1}:00 => {program or 'Free'}")

File: QUESTION2/question.js

const programs = [
  { name: "News", duration: 1, reach: 9, revenue: 1000, rights: [0, 24] },
  { name: "Movie", duration: 2, reach: 10, revenue: 5000, rights: [18, 22] },
  { name: "Kids Show", duration: 1, reach: 6, revenue: 800, rights: [8, 18] }
];

const slots = new Array(24).fill(null);
const primeTime = [18, 19, 20, 21];

programs.sort((a, b) => b.reach * b.revenue - a.reach * a.revenue);

for (const program of programs) {
  const [start, end] = program.rights;
  for (let i = start; i <= end - program.duration; i++) {
    let canFit = true;
    for (let j = i; j < i + program.duration; j++) {
      if (slots[j]) {
        canFit = false;
        break;
      }
    }

    if (canFit) {
      const isPrime = primeTime.some(t => t >= i && t < i + program.duration);
      if (isPrime || program.reach >= 8) {
        for (let j = i; j < i + program.duration; j++) {
          slots[j] = program.name;
        }
        break;
      }
    }
  }
}

// print schedule
slots.forEach((program, hour) => {
  console.log(`${hour}:00 - ${hour + 1}:00 => ${program || "Free"}`);
});